### ISSUE TITLE
INSERT INTO SELECT between 2 tables with rep_factor 2 and with primary keys shows wrong result

### ISSUE THREAD

**Issue by saicitus** [citusdata/citus#2209]
Unfortunately I don't have the exact steps for reproducing the issue. Below are **the pseudo steps** of how this happened in a customer cluster. The source table had close to 4.5 billion rows.

**Source Table Setup**
```
CREATE TABLE test(id serial, tenant_id int, primary key (id,tenant_id));
SELECT create_distributed_table('test','tenant_id');
--Load data.
SET citus.shard_replication_factor to 2;
SELECT replicate_table_shards('test');
```
 **Destination Table Setup**
```
--Creating a new table with replication_factor setting 1 is not respected i.e it's replication_factor is 2 in real
show citus.shard_replication_factor ;
 citus.shard_replication_factor
--------------------------------
 1
(1 row)
CREATE TABLE dest(id serial, tenant_id int, primary key (id,tenant_id));
SELECT create_distributed_table('dest','tenant_id');
SELECT count(*),shardid from pg_dist_shard_placement JOIN pg_dist_shard USING (shardid) where logicalrelid='dest'::regclass group by shardid;
 count | shardid
-------+---------
     2 |  102136
     2 |  102137
     2 |  102138
     2 |  102139
     2 |  102140
     2 |  102141
     2 |  102142
     2 |  102143
...
...

SELECT colocationid,logicalrelid from pg_dist_partition;
 colocationid |  logicalrelid
--------------+-----------------
            1 | test
            1 |dest
```
**INSERT INTO SELECT shows wrong results**
```
INSERT INTO dest SELECT * from test;
WARNING: (1233,2) violates pkey constraint etc etc.
DETAIL: etc etcc.
WARNING: (1233,2) violates pkey constraint etc etc.
DETAIL: etc etcc.
WARNING: (1233,2) violates pkey constraint etc etc.
DETAIL: etc etcc.

SELECT count(*) from test;
count
----
4.6 billion
SELECT count(*) from dest;
count
---
4.4 billion
```
Instead of WARNING we should have errored out.
Im not sure why this issue happened, it could also be because of 'id' serial column in the destination table. And Im not sure whether the id column in the source table was populated.

**Comment by onderkalaci** [comment]
I cannot reproduce that, any failure should always fail the entire distributed transaction. (Techinal note: The connection is always marked as critical via  `MarkRemoteTransactionCritical()` for the last 3-4 Citus relases that I've checked.)


```SQL
ERROR:  duplicate key value violates unique constraint "dest_pkey_102139"
DETAIL:  Key (id, tenant_id)=(1, 1) already exists.
CONTEXT:  while executing command on localhost:9701
```

I've some ideas on how this might be happening. Could you please reply the following questions?

1) Is the `INSERT .. SELECT` failing? Or is that successfully committing with `WARNING`s? If it is not committing successfully, should we classify this issue as a bug?
2) Which Citus version are you running on?
3)  Is `MultiShardCommitProtocol = 2PC`?
4) Have you done a single `INSERT ... SELECT` for the whole `4.6b` rows, or, is it a gradual operation where `dest` is ingested regularly (probably with filters on the distribution key)?
5) Are there any `INVALID` shard placement states either in the source or dest?

> And Im not sure whether the id column in the source table was populated.

I'm assuming the answer is yes because the column is in the primary key.


> Im not sure why this issue happened, it could also be because of 'id' serial column in the destination table. And Im not sure whether the id column in the source table was populated.

Although I cannot reproduce it, the issue sounds more related to the serial column. As you might already know, serial columns are associated with `sequences` and `sequences` are not transactional (e.g., any `rollback` doesn't decrement the value).

```SQL
CREATE TABLE test(id serial, tenant_id int, primary key (id,tenant_id));
CREATE TABLE
INSERT INTO test (tenant_id) VALUES (1);
INSERT 0 1
INSERT INTO test (tenant_id) VALUES (1);
INSERT 0 1
SELECT * FROM test;
 id | tenant_id 
----+-----------
  1 |         1
  2 |         1
(2 rows)

BEGIN;
INSERT INTO test (tenant_id) VALUES (1);
INSERT 0 1
ROLLBACK;
INSERT INTO test (tenant_id) VALUES (1);
INSERT 0 1
-- note that the id is 4 
SELECT * FROM test;
 id | tenant_id 
----+-----------
  1 |         1
  2 |         1
  4 |         1
```

I'll try more to replicate the problem if you can answer my above questions. Because, otherwise the search space to spot the issue is huge.

**Comment by saicitus** [comment]
>Is the INSERT .. SELECT failing? Or is that successfully committing with WARNINGs? If it is not committing successfully, should we classify this issue as a bug?

Yes it is successfully committing with WARNINGs. I guess it is a bug then.

>Which Citus version are you running on?

Citus Enterprise 7.3

>Is MultiShardCommitProtocol = 2PC?

I think it is 2PC because I believe the customer has changed the default.

>Have you done a single INSERT ... SELECT for the whole 4.6b rows, or, is it a gradual operation where dest is ingested regularly (probably with filters on the distribution key)?

Single INSERT INTO SELECT

>Are there any INVALID shard placement states either in the source or dest?

I haven't checked, so Im not sure.

> I'm assuming the answer is yes because the column is in the primary key.

You are right 👍 

>Although I cannot reproduce it, the issue sounds more related to the serial column. As you might already know, serial columns are associated with sequences and sequences are not transactional (e.g., any rollback doesn't decrement the value).

Yes that makes sense.
But I'd assume the sequence never to be used when the source table already the id column populated. I have already recommended to the customer to stash the serial column on the destination table just as a best practice.
